### PR TITLE
Follow-up requests

### PR DESCRIPTION
**Pull request by acrellin**
Sample available parameters 1:
followup_requests

Sample available parameters 2:
followup_requests_2

### CODE REVIEW

**Comment by acrellin** [comment]
[ch759]

**Comment by acrellin** [comment]
[ch758]

**Comment by acrellin** [comment]
@stefanv this still needs tests and an `editRequest` handler, but 99% of the code is in now. If you could please review this in the next week, I will add the remaining pieces in a separate commit that you can review independently afterwards.

File: instrument_observation_params.json [hunk 1]
```diff
@@ -0,0 +1,21 @@
+{
```
Comment (acrellin) [inline instrument_observation_params.json:1]: Alternatively, this could live in a JSONB column in the instruments table.

**Comment by dannygoldstein** [comment]
Is there a way to use this to request spectroscopy, e.g., from SEDM on the p60?

**Comment by acrellin** [comment]
We could easily extend this to add spectroscopic followup requests

File: instrument_observation_params.json (hunk 1 above)
Comment (profjsb) [reply instrument_observation_params.json:1]: Perhaps check to see if this file exists or not (or is empty). In which case, this part of the Source page is not rendered.

File: skyportal/handlers/api/followup_request.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+                          type: string
+                          description: New follow-up request ID
+        """
+        data = self.get_json()
+        _ = Source.get_if_owned_by(data["source_id"], self.current_user)
+        data["start_date"] = arrow.get(data["start_date"]).datetime
+        data["end_date"] = arrow.get(data["end_date"]).datetime
+        data["requester_id"] = self.current_user.id
+        if isinstance(data["filters"], str):
+            data["filters"] = [data["filters"]]
```
Comment (profjsb) [inline skyportal/handlers/api/followup_request.py:38]: Should be catching errors here -- for instance if the `start_date` is improperly formatted (ie. arrow throws an error)

File: static/js/components/FollowupRequestForm.jsx [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+        };
+      }
+    }
+    setFormState({
+      ...formState,
+      ...newState
+    });
+  };
+
+  const handleSubmit = () => {
```
Comment (profjsb) [inline static/js/components/FollowupRequestForm.jsx:61]: here's where you can validate the dates on the client side (e.g. https://www.telerik.com/kendo-react-ui/components/dateinputs/datepicker/forms/#toc-setting-validation-requirements)

File: skyportal/handlers/api/followup_request.py (hunk 2 above)
Comment (acrellin) [reply skyportal/handlers/api/followup_request.py:38]: @profjsb if the parse fails, it automatically returns with an error code & message such that a notification is displayed on the front-end with the arrow exception message: `Backend error: Could not match input 'mydate' to any of the following formats: YYYY-MM-DD, YYYY-M-DD, YYYY-M-D, YYYY/MM/DD, YYYY/M/DD, YYYY/M/D, YYYY.MM.DD, YYYY.M.DD, YYYY.M.D, YYYYMMDD, YYYY-DDDD, YYYYDDDD, YYYY-MM, YYYY/MM, YYYY.MM, YYYY`. Does that suffice?

File: instrument_observation_params.json (hunk 1 above)
Comment (acrellin) [reply instrument_observation_params.json:1]: Done (see the handler method)

File: static/js/components/FollowupRequestForm.jsx (hunk 3 above)
Comment (acrellin) [reply static/js/components/FollowupRequestForm.jsx:61]: How does this look? https://reactdatepicker.com/

File: skyportal/handlers/api/followup_request.py (hunk 2 above)
Comment (profjsb) [reply skyportal/handlers/api/followup_request.py:38]: Sure.

File: static/js/components/FollowupRequestForm.jsx (hunk 3 above)
Comment (profjsb) [reply static/js/components/FollowupRequestForm.jsx:61]: Yes, good. For a picker this is fine. But does it also validate on the client side?

Comment (acrellin) [reply static/js/components/FollowupRequestForm.jsx:64]: It doesn't allow for invalid dates - if you enter some invalid text, upon element blur, it reverts back to the last valid input.

Comment (acrellin) [reply static/js/components/FollowupRequestForm.jsx:64]: Done.

**Comment by acrellin** [comment]
Updated UI screenshot to show date picker: 
followup_requests_form

**Comment by acrellin** [comment]
[ch864]

**Comment by acrellin** [comment]
Now includes an edit button which, when clicked, displays a modal with a form to edit the follow-up request. Here it is in action (note that the relevant row is automatically updated via web-socket message & re-fetching of data asynchronously): 


edit_followup_request_form

**Comment by acrellin** [comment]
[ch873]

File: skyportal/handlers/api/followup_request.py (hunk 2 above)
Comment (stefanv) [reply skyportal/handlers/api/followup_request.py:38]: As noted in the meeting earlier today, we should catch all errors so they don't get thrown to the backend logs.

File: skyportal/handlers/api/followup_request.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+                    - Success
+                    - type: object
+                      properties:
+                        id:
+                          type: string
+                          description: New follow-up request ID
+        """
+        data = self.get_json()
+        _ = Source.get_if_owned_by(data["source_id"], self.current_user)
+        data["start_date"] = arrow.get(data["start_date"]).datetime
```
Comment (stefanv) [inline skyportal/handlers/api/followup_request.py:34]: The arguments should be documented.

File: skyportal/handlers/api/internal/instrument_observation_params.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+
+class InstrumentObservationParamsHandler(BaseHandler):
+    @auth_or_token
+    def get(self):
+        try:
+            with open("instrument_observation_params.json") as f:
+                instrument_data = json.load(f)
+        except FileNotFoundError:
+            return self.error("instrument_observation_params.json does not exist.")
```
Comment (stefanv) [inline skyportal/handlers/api/internal/instrument_observation_params.py:13]: This leaks implementation details to the user.  How about something like: "Instrument observation parameters unavailable."

File: skyportal/handlers/api/internal/instrument_observation_params.py [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+        except json.JSONDecodeError:
+            return self.error("JSON parse error: instrument_observation_params.json "
```
Comment (stefanv) [inline skyportal/handlers/api/internal/instrument_observation_params.py:15]: Same here:

"System misconfiguration: invalid instrument observation parameters."

File: static/js/components/EditFollowupRequestDialog.jsx [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    <span key={followupRequest.id}>
+      <button type="button" onClick={handleClickOpen}>
+        Edit
+      </button>
+      <Dialog open={open} onClose={handleClose} style={{ position: "fixed" }}>
+        <DialogContent>
+          <FollowupRequestForm
+            source_id={followupRequest.source_id}
+            action="editExisting"
+            followupRequest={followupRequest}
```
Comment (stefanv) [inline static/js/components/EditFollowupRequestDialog.jsx:30]: You should be able to unpack these automatically without specifying each field.

File: static/js/components/FollowupRequestForm.jsx [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+import PropTypes from 'prop-types';
+import DatePicker from 'react-datepicker';
+
+import { showNotification } from 'baselayer/components/Notifications';
+import * as Actions from '../ducks/source';
+
+import 'react-datepicker/dist/react-datepicker-cssmodules.css';
+
+
+const FollowupRequestForm = ({ source_id, action, instrumentList, instrumentObsParams, followupRequest = null, title = "Submit new follow-up request", afterSubmit = null }) => {
```
Comment (stefanv) [inline static/js/components/FollowupRequestForm.jsx:12]: This form could probably benefit quite a bit from React Hook Form!  Will let you investigate and see if you agree!

File: webpack.config.js [hunk 9]
```diff
@@ -52,7 +52,7 @@ const config = {
       // Enable CSS Modules for Skyportal
       {
         test: /\.css$/,
-        include: /static\/js/,
+        include: [/static\/js/, /node_modules\/react-datepicker\/dist/],
```
Comment (stefanv) [inline webpack.config.js:55]: Why is this necessary?

File: skyportal/handlers/api/followup_request.py (hunk 4 above)
Comment (acrellin) [reply skyportal/handlers/api/followup_request.py:34]: We specify the schema, as in our other handlers, so these fields will automatically show up in the API docs

File: skyportal/handlers/api/internal/instrument_observation_params.py (hunk 5 above)
Comment (acrellin) [reply skyportal/handlers/api/internal/instrument_observation_params.py:13]: Sure, done.

File: static/js/components/EditFollowupRequestDialog.jsx (hunk 7 above)
Comment (acrellin) [reply static/js/components/EditFollowupRequestDialog.jsx:30]: Can you please explain? This is just standard props-passing

File: static/js/components/FollowupRequestForm.jsx (hunk 8 above)
Comment (acrellin) [reply static/js/components/FollowupRequestForm.jsx:12]: Will investigate...

File: webpack.config.js (hunk 9 above)
Comment (acrellin) [reply webpack.config.js:55]: This is needed for this import line: `import 'react-datepicker/dist/react-datepicker-cssmodules.css';`

File: skyportal/handlers/api/followup_request.py (hunk 4 above)
Comment (stefanv) [reply skyportal/handlers/api/followup_request.py:34]: The input fields do not show up automatically for get requests, do they?

Comment (acrellin) [reply skyportal/handlers/api/followup_request.py:34]: this is a POST handler...

Comment (acrellin) [reply skyportal/handlers/api/followup_request.py:34]: Look at any other POST handler -- we always use the schema to create the API docs. Look above here in this file:
```
        requestBody:
          content:
            application/json:
              schema: FollowupRequest
```

Comment (stefanv) [reply skyportal/handlers/api/followup_request.py:34]: You're right; thanks.

File: static/js/components/EditFollowupRequestDialog.jsx (hunk 7 above)
Comment (stefanv) [reply static/js/components/EditFollowupRequestDialog.jsx:30]:
```
<ChildComponent {...props} />
```

Comment (acrellin) [reply static/js/components/EditFollowupRequestDialog.jsx:30]: Only 3 of the 7 props being passed into `<ChildComponent>` are also props of `<ParentComponent>` (`followupRequest, instrumentList, instrumentObsParams`), and since we follow the object destructuring pattern for props in our functional component definitions, this isn't available here.

Comment (acrellin) [reply static/js/components/EditFollowupRequestDialog.jsx]: Sure, I changed the function signature to use a single props object and not use destructuring so that we can do `{...props}` here.

Comment (acrellin) [reply static/js/components/EditFollowupRequestDialog.jsx]: Actually, that's an anti-pattern - reverting to what I had done originally which is the correct form. 
```
/home/travis/build/skyportal/skyportal/static/js/components/EditFollowupRequestDialog.jsx
  21:16  error  Must use destructuring props assignment  react/destructuring-assignment
  28:24  error  Must use destructuring props assignment  react/destructuring-assignment
  32:13  error  Prop spreading is forbidden              react/jsx-props-no-spreading
```

Comment (stefanv) [reply static/js/components/EditFollowupRequestDialog.jsx:30]: Ooh forbidden! :fire:

**Comment by stefanv** [comment]
Loving the new hook-form integration—thanks for figuring that out for us, Ari!

In it goes :rocket: !

**Comment by dannygoldstein** [comment]
Hi guys, a few minor comments - 

  * Exposure time should be a free form text entry field with a default value and min and max values for each instrument

Details - 

  * P60 Rainbow Camera only has sdssu, sdssg, sdssr, and sdssi filters

  * ALFOSC has Bessell UBVRI and some other filters, I think @arielgoobar knows the full filter set
